nixos/modules/virtualisation/amazon-image.nix

```diff
@@ -66,6 +68,32 @@ in
     boot.loader.timeout = 0;
 
     boot.initrd.network.enable = true;
+    boot.initrd.postDeviceCommands = lib.mkMerge [
+      (lib.mkBefore ''
+        (
+          set -x
+          PATH=${pkgs.jq}/bin/:${pkgs.cloud-utils}/bin/:$PATH
+          lsblk --json -O --all \
+            | jq -cf ${./amazon-image-zfs.jq} \
+            | (while read -r line; do
+                part_path=$(echo "$line" | jq -r .part_path);
+                disk_path=$(echo "$line" | jq -r .disk_path);
+                growpart "$disk_path" "$(cat "$part_path")";
+              done)
+        )
+      '')
+      # zfs mounts within postDeviceCommands so mkBefore and mkAfter must be used
+      (lib.mkAfter ''
+        (
+          set -x
+          for pool in $(zpool list -H | awk '{print $1}'); do
+            for vdev in $(zpool list -vLPH "$pool" | awk '($1 ~ "^/dev" && $10 = "ONLINE") {print $1;}'); do
+              zpool online -e "$pool" "$vdev";
+            done
+          done
+        )
+      '')
```

Comment:
These two bits are a bit weird. To resize the disk on startup we need to manually grow the partitions and then `zpool online -e` the devices. The second part wouldn't be necessary if our ZFS's Zed properly handled autoexpand, but for some reason I did not see that working and manually implemented it.

Reply:
Learning today that these should not be included in the image if ZFS is not enabled. They don't appear to cause problems, but it isn't necessary.